### PR TITLE
DBZ-6970 Debezium jdbc sink process truncate event failure

### PR DESCRIPTION
**Pull request by harveyyue**
https://issues.redhat.com/browse/DBZ-6970

### CODE REVIEW

File: src/main/java/io/debezium/connector/jdbc/JdbcSinkConnectorConfig.java [hunk 1]
```diff
@@ -150,6 +151,16 @@ public class JdbcSinkConnectorConfig {
             .withValidation(JdbcSinkConnectorConfig::validateDeleteEnabled)
             .withDescription("Whether to treat `null` record values as deletes. Requires primary.key.mode to be `record.key`.");
 
+    public static final Field TRUNCATE_ENABLED_FIELD = Field.create(TRUNCATE_ENABLED)
+            .withDisplayName("Controls whether records can be truncated by the connector")
```
Comment (mfvitale) [inline src/main/java/io/debezium/connector/jdbc/JdbcSinkConnectorConfig.java:155]:
```suggestion
            .withDisplayName("Controls whether tables can be truncated by the connector")
```

File: src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java [hunk 2]
```diff
@@ -282,8 +290,9 @@ public SinkRecordDescriptor build() {
             Objects.requireNonNull(sinkRecord, "The sink record must be provided.");
 
             final boolean flattened = !isTombstone(sinkRecord) && isFlattened(sinkRecord);
-            readSinkRecordKeyData(sinkRecord, flattened);
-            readSinkRecordNonKeyData(sinkRecord, flattened);
+            final boolean truncated = !flattened && isTruncateEvent(sinkRecord);
+            readSinkRecordKeyData(sinkRecord, flattened, truncated);
```
Comment (mfvitale) [inline src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:294]: What do you think to the check here about the `truncated` and remove it from the `readSinkRecord*` methods?

**Comment by mfvitale** [comment]
Hi @harveyyue generally LGTM, I just leaved small suggestions.

**Comment by harveyyue** [comment]
> Hi @harveyyue generally LGTM, I just leaved small suggestions. Also can you please check to the DB2 test failure? Seems related to the changes.

Done, it's seams fairure with executing truncate statement, I will double check later.

**Review by mfvitale — APPROVED** [review]

**Comment by harveyyue** [comment]
@mfvitale The truncate statement cause the db2 it test failure, and I'm trying the "truncate table xxx" and "[truncate table xxx immediate](https://troels.arvin.dk/db/rdbms/#bulk-truncate_table-db2)" statements are not work. Unfortunatly, I couldn't setup a db2 instance via docker in my laptop mac m1, if you have a db2 env, please help to verify this truncate statement or the user account privileges.

`
Caused by: jakarta.persistence.PersistenceException: Converting `org.hibernate.exception.GenericJDBCException` to JPA `PersistenceException` : JDBC exception executing SQL [TRUNCATE TABLE DB2INST1.SERVER1_SCHEMA_HJE9VZB IMMEDIATE]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:165)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:175)
	at org.hibernate.query.spi.AbstractQuery.executeUpdate(AbstractQuery.java:654)
	at io.debezium.connector.jdbc.JdbcChangeEventSink.writeTruncate(JdbcChangeEventSink.java:331)
	at io.debezium.connector.jdbc.JdbcChangeEventSink.write(JdbcChangeEventSink.java:226)
	at io.debezium.connector.jdbc.JdbcChangeEventSink.execute(JdbcChangeEventSink.java:84)
	... 119 common frames omitted
Caused by: org.hibernate.exception.GenericJDBCException: JDBC exception executing SQL [TRUNCATE TABLE DB2INST1.SERVER1_SCHEMA_HJE9VZB IMMEDIATE]
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:61)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:109)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:95)
	at org.hibernate.sql.exec.internal.StandardJdbcMutationExecutor.execute(StandardJdbcMutationExecutor.java:97)
	at org.hibernate.query.sql.internal.NativeNonSelectQueryPlanImpl.executeUpdate(NativeNonSelectQueryPlanImpl.java:78)
	at org.hibernate.query.sql.internal.NativeQueryImpl.doExecuteUpdate(NativeQueryImpl.java:820)
	at org.hibernate.query.spi.AbstractQuery.executeUpdate(AbstractQuery.java:643)
	... 122 common frames omitted
Caused by: com.ibm.db2.jcc.am.SqlException: DB2 SQL Error: SQLCODE=-428, SQLSTATE=25001, SQLERRMC=null, DRIVER=4.26.14
	at com.ibm.db2.jcc.am.b7.a(b7.java:815)
	at com.ibm.db2.jcc.am.b7.a(b7.java:66)
	at com.ibm.db2.jcc.am.b7.a(b7.java:140)
	at com.ibm.db2.jcc.am.k4.b(k4.java:2471)
	at com.ibm.db2.jcc.am.k4.c(k4.java:2452)
	at com.ibm.db2.jcc.t4.ab.n(ab.java:914)
	at com.ibm.db2.jcc.t4.ab.a(ab.java:120)
	at com.ibm.db2.jcc.t4.p.a(p.java:50)
	at com.ibm.db2.jcc.t4.aw.b(aw.java:220)
	at com.ibm.db2.jcc.am.k5.bm(k5.java:3599)
	at com.ibm.db2.jcc.am.k5.a(k5.java:4644)
	at com.ibm.db2.jcc.am.k5.b(k5.java:4182)
	at com.ibm.db2.jcc.am.k5.be(k5.java:827)
	at com.ibm.db2.jcc.am.k5.executeUpdate(k5.java:801)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:1502)
	at org.hibernate.sql.exec.internal.StandardJdbcMutationExecutor.execute(StandardJdbcMutationExecutor.java:84)
	... 125 common frames omitted
`

**Comment by mfvitale** [comment]
> @mfvitale The truncate statement cause the db2 it test failure, and I'm trying the "truncate table xxx" and "[truncate table xxx immediate](https://troels.arvin.dk/db/rdbms/#bulk-truncate_table-db2)" statements are not work. Unfortunatly, I couldn't setup a db2 instance via docker in my laptop mac m1, if you have a db2 env, please help to verify this truncate statement or the user account privileges.
> 
> `Caused by: jakarta.persistence.PersistenceException: Converting`org.hibernate.exception.GenericJDBCException`to JPA`PersistenceException`: JDBC exception executing SQL [TRUNCATE TABLE DB2INST1.SERVER1_SCHEMA_HJE9VZB IMMEDIATE] at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:165) at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:175) at org.hibernate.query.spi.AbstractQuery.executeUpdate(AbstractQuery.java:654) at io.debezium.connector.jdbc.JdbcChangeEventSink.writeTruncate(JdbcChangeEventSink.java:331) at io.debezium.connector.jdbc.JdbcChangeEventSink.write(JdbcChangeEventSink.java:226) at io.debezium.connector.jdbc.JdbcChangeEventSink.execute(JdbcChangeEventSink.java:84) ... 119 common frames omitted Caused by: org.hibernate.exception.GenericJDBCException: JDBC exception executing SQL [TRUNCATE TABLE DB2INST1.SERVER1_SCHEMA_HJE9VZB IMMEDIATE] at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:61) at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:109) at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:95) at org.hibernate.sql.exec.internal.StandardJdbcMutationExecutor.execute(StandardJdbcMutationExecutor.java:97) at org.hibernate.query.sql.internal.NativeNonSelectQueryPlanImpl.executeUpdate(NativeNonSelectQueryPlanImpl.java:78) at org.hibernate.query.sql.internal.NativeQueryImpl.doExecuteUpdate(NativeQueryImpl.java:820) at org.hibernate.query.spi.AbstractQuery.executeUpdate(AbstractQuery.java:643) ... 122 common frames omitted Caused by: com.ibm.db2.jcc.am.SqlException: DB2 SQL Error: SQLCODE=-428, SQLSTATE=25001, SQLERRMC=null, DRIVER=4.26.14 at com.ibm.db2.jcc.am.b7.a(b7.java:815) at com.ibm.db2.jcc.am.b7.a(b7.java:66) at com.ibm.db2.jcc.am.b7.a(b7.java:140) at com.ibm.db2.jcc.am.k4.b(k4.java:2471) at com.ibm.db2.jcc.am.k4.c(k4.java:2452) at com.ibm.db2.jcc.t4.ab.n(ab.java:914) at com.ibm.db2.jcc.t4.ab.a(ab.java:120) at com.ibm.db2.jcc.t4.p.a(p.java:50) at com.ibm.db2.jcc.t4.aw.b(aw.java:220) at com.ibm.db2.jcc.am.k5.bm(k5.java:3599) at com.ibm.db2.jcc.am.k5.a(k5.java:4644) at com.ibm.db2.jcc.am.k5.b(k5.java:4182) at com.ibm.db2.jcc.am.k5.be(k5.java:827) at com.ibm.db2.jcc.am.k5.executeUpdate(k5.java:801) at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:1502) at org.hibernate.sql.exec.internal.StandardJdbcMutationExecutor.execute(StandardJdbcMutationExecutor.java:84) ... 125 common frames omitted`

Hi @harveyyue , 

I have tried with 
```sql
TRUNCATE TABLE PRODUCTS 
```
and got 

```text
SQL Error [42601]: An unexpected token "END-OF-STATEMENT" was found following "NCATE TABLE PRODUCTS".  Expected tokens may include:  "IMMEDIATE".. SQLCODE=-104, SQLSTATE=42601, DRIVER=4.31.10
```

with the `IMMEDIATE` instead I got not error but the table is still there. Tested with DB2  v11.5.8.0 from ibmcom/db2 (same used by tests). 

Let me do more investigation.

**Comment by mfvitale** [comment]
@harveyyue I have fixed your test with plain JDBC and passes on local setup. At this point I suppose that the issue is with Hibernate as reported by @Naros [here](https://debezium.zulipchat.com/#narrow/stream/348103-community-db2/topic/Truncate.20statement). Can you investigate more?

**Comment by harveyyue** [comment]
@mfvitale THX😊

**Comment by mfvitale** [comment]
I have investigated it more and this is my finding:
* Removing c3p0 dependency didn't worked. 
* Enabling Hibernates logs didn't showed any useful information to prove the hypothesis that Hibernates adds some statements hiddenly that causes the `TRUNCATE` to fail when is not the first statement in the transaction. 

So the only way to let it working is to use the `ALTER TABLE "SCHEMA"."TABLENAME" activate not logged initially with empty table` statement. 

@Naros WDYT?

**Comment by Naros** [comment]
Let me quickly look at something and get back to you @mfvitale.  

My only concern is that every database is slightly different when it comes to `TRUNCATE` support.  For example, in Db2 you either need `DELETE`, `CONTROL`, or `DATACCESS` permissions but with the `ALTER TABLE` the user must either have both `ALTER` and `DELETE` privileges, `CONTROL`, `SCHEMAADM`, or `DBADM`.  

If my findings turn up nil, then we'll just document this requirement for Db2 and move on.

**Comment by mfvitale** [comment]
@harveyyue  Applied, thanks!

**Pull request merged**